src/dsp_tools/utils/xml_upload_enhanced.py

```diff
@@ -0,0 +1,113 @@
+from typing import Tuple
+import requests
+from itertools import repeat
+import concurrent.futures
+from pathlib import Path
+import os
+
+
+def generate_testdata() -> None:
+    """
+    Creates a test data folder in the user's current working directory.
+
+    Returns:
+        None
+    """
+    testproject = Path(os.getcwd()) / "enhanced-xmlupload-testproject"
+    if testproject.exists():
+        print("The test project folder is already existing.")
+        return
+    destinations = [
+        testproject / "multimedia",
+        testproject / "multimedia" / "nested",
+        testproject / "multimedia" / "nested" / "subfolder"
+    ]
+    for sub in destinations:
+        sub.mkdir(parents=True)
+    github_bitstreams_path = "https://github.com/dasch-swiss/dsp-tools/blob/main/testdata/bitstreams"
+    ext_img = ["jpg", "jpeg", "tif", "tiff", "jp2", "png"]
+    for ext in ext_img:
+        img = requests.get(f"{github_bitstreams_path}/test.{ext}?raw=true").content
+        for dst in destinations:
+            with open(dst / f"test.{ext}", "bw") as f:
+                f.write(img)
+    print(f"Successfully created folder {testproject}")
+
+    # TODO: generate an XML file that uses these files
+
+
+def check_multimedia_folder(
+    xmlfile: str,
+    multimedia_folder: str
+) -> bool:
+    """
+    Verify that all multimedia files referenced in the XML file are contained in multimedia_folder
+    (or one of its subfolders), and that all files contained in multimedia_folder are referenced in the XML file.
+    """
+
+    # TODO
+    return True
+
+
+def make_batches(multimedia_folder: str) -> Tuple[list[list[Path]], int]:
+    """
+    Make 32 portions (batches) out of all images in the multimedia folder, one batch per thread.
+    A batch is a list of file paths.
+
+    Returns:
+        batches, batch size
+    """
+    # TODO: Vij is working on this. This is only a mock-up
+    batches = [
+        [x for x in Path(multimedia_folder).iterdir() if x.is_file()],
+        [x for x in Path(multimedia_folder + "/nested").iterdir() if x.is_file()],
+        [x for x in Path(multimedia_folder + "/nested/subfolder").iterdir() if x.is_file()]
+    ]
+    batch_size = 6
+    return batches, batch_size
+
+
+def process_seq(batch: list[Path], sipi_port: int, id: int) -> None:
+    """
+    Upload the images contained in one batch, sequentially.
+
+    Args:
+        batch: list of paths to image files
+        sipi_port: port number of SIPI
+
+    Returns:
+        None
+    """
+    print(f"Process batch with ID {id}...")
+    for imgpath in batch:
+        response = requests.post(url=f'http://localhost:{sipi_port}/upload', files={'file': open(imgpath, 'rb')})
```

Comment:
@irinaschubert Perhaps you can help here. I always get a 400 error "Token missing". Vij says that after commenting out the passage in `sipi/scripts/upload.lua`, this error shouldn't happen any more, but it persists. To see what we comment out in `sipi/scripts/upload.lua`, see in this PR the file `xmlupload-enhanced.md`, line 51

Reply:
I found it out. I had a "dsp-api" repo on my desktop that I left untouched, and a "dsp-api-modified" where I modified the contents. But when starting the Docker container, it took some data and configuration from the "dsp-api" folder. Now that I renamed these two folders, it works.

Reply:
cool :)